### PR TITLE
Malicious site protection - extract special error page navigation logic

### PR DESCRIPTION
**Pull request by alessandroboron**
Task/Issue URL: https://app.asana.com/0/1206329551987282/1208836682251611/f
Tech Design URL: https://app.asana.com/0/1206329551987282/1207273224076495/f

**Description**:

This PR extracts the navigation logic for special error pages in its own class. 

This PR focuses on creating the the entites used to encapsulate the logic for SSL and Malicious site protection

The implementation of the logic for Malicious site protection will follow in an upcoming PR.

**Steps to test this PR**:
1. Ensure all the SpecialErrorPage tests pass

**Manual tests**
SSL Tests from [SSL PR](https://github.com/duckduckgo/iOS/pull/3271)

**Steps to test this PR**:
1. Go to badssl.com
2. Select expired certificate
3. Our custom special error page should open. 
- at this point you should be able to go back, and once you go back you should be able to go forward
- please validate if design looks fine, and if it doesn't break in landscape
- please validate if design supports light/dark modes
- please validate if design supports translations (in order to test it you will most likely have to re-enter the page)
- please validate the privacy shield icon in navigation bar is invisible.
4. Tap on leave site.
- you should be taken back (if it was possible) or the tab should be closed if this was first page in navigation
5. Tap on advanced and accept the risk.
- please validate that expired.badssl.com has loaded (red background)
- please validate that privacy shield icon reappeared but has red dot on top of it
- once you go to privacy dashboard you should see information about incorrect certificate
- please validate you can go back to badssl.com site

**Definition of Done (Internal Only)**:

* [x] Does this PR satisfy our [Definition of Done](https://app.asana.com/0/1202500774821704/1207634633537039/f)?

**Copy Testing**:

* [ ] Use of correct apostrophes in new copy, ie `’` rather than `'`

**Orientation Testing**:

* [ ] Portrait
* [ ] Landscape

**Device Testing**:

* [ ] iPhone SE (1st Gen)
* [ ] iPhone 8
* [ ] iPhone X
* [ ] iPhone 14 Pro
* [ ] iPad

**OS Testing**:

* [ ] iOS 15
* [ ] iOS 16
* [ ] iOS 17

**Theme Testing**:

* [ ] Light theme
* [ ] Dark theme

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

File: DuckDuckGo/MaliciousSiteProtection/MaliciousSiteProtectionManager.swift [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+
+final class MaliciousSiteProtectionManager: MaliciousSiteDetecting {
```
Comment (alessandroboron) [inline DuckDuckGo/MaliciousSiteProtection/MaliciousSiteProtectionManager.swift:22]: Implementation of this class will come in a subsequent PR.
This was used to mock the BSK library and assert that the special error page was rendered.

File: DuckDuckGo/SpecialErrorPage/SpecialErrorPageNavigationHandler+MaliciousSite.swift [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+//  limitations under the License.
+//
+
+import Foundation
+import BrowserServicesKit
+import Core
+import SpecialErrorPages
+import WebKit
+
+enum MaliciousSiteProtectionNavigationResult: Equatable {
```
Comment (alessandroboron) [inline DuckDuckGo/SpecialErrorPage/SpecialErrorPageNavigationHandler+MaliciousSite.swift:26]: This class will encapsulate MaliciousSiteProtection logic to evaluate if a site is malicious and will handle its internal state to bypass URLs.

File: DuckDuckGo/SpecialErrorPage/SpecialErrorPageNavigationHandler+SSL.swift [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        let errorData = SpecialErrorData(kind: .ssl,
+                                         errorType: errorType.rawValue,
+                                         domain: failedURL.host,
+                                         eTldPlus1: storageCache.tld.eTLDplus1(failedURL.host))
+
+        return SSLSpecialError(url: failedURL, type: errorType, errorData: errorData)
+    }
+
+    func errorPageVisited(errorType: SSLErrorType) {
+        Pixel.fire(pixel: .certificateWarningDisplayed(errorType.rawParameter))
```
Comment (alessandroboron) [inline DuckDuckGo/SpecialErrorPage/SpecialErrorPageNavigationHandler+SSL.swift:82]: We could inject the fire function and add tests to ensure Pixels are fired.

File: DuckDuckGo/SpecialErrorPage/SpecialErrorPageNavigationHandler+SSL.swift [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+import Common
+import BrowserServicesKit
+import SpecialErrorPages
+import Core
+
```
Comment (alessandroboron) [inline DuckDuckGo/SpecialErrorPage/SpecialErrorPageNavigationHandler+SSL.swift:25]: This class encapsulate the SSL logic

File: DuckDuckGo/SpecialErrorPage/SpecialErrorPageNavigationHandler.swift [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+import WebKit
+import SpecialErrorPages
+import Core
+
```
Comment (alessandroboron) [inline DuckDuckGo/SpecialErrorPage/SpecialErrorPageNavigationHandler.swift:24]: This class will orchestrate the navigation logic of features that require a Special Error Page

File: DuckDuckGoTests/SpecialErrorPage/SSLErrorPageNavigationHandlerTests.swift [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+//
+
+import Foundation
+import XCTest
+import WebKit
+
+@testable import SpecialErrorPages
+@testable import DuckDuckGo
+
+final class SSLSpecialErrorPageTests: XCTestCase {
```
Comment (alessandroboron) [inline DuckDuckGoTests/SpecialErrorPage/SSLErrorPageNavigationHandlerTests.swift]: These tests were initially using the TabViewController. I split them in two test classes, `SSLSpecialErrorPageTests` and `SpecialErrorPageNavigationHandlerIntegrationTests`.

We could also convert this class using SwiftTesting so we have all test suites related to Special Error Pages. I can make a follow up task at the end of the project.

File: DuckDuckGo/SpecialErrorPage/SpecialErrorPageNavigationHandler+SSL.swift [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+protocol SSLSpecialErrorPageNavigationHandling {
+    func handleServerTrustChallenge(_ challenge: URLAuthenticationChallenge, completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void)
+
+    func makeNewRequestURLAndSpecialErrorDataIfEnabled(error: NSError) -> SSLSpecialError?
+
+    func errorPageVisited(errorType: SSLErrorType)
```
Comment (jaceklyp) [inline DuckDuckGo/SpecialErrorPage/SpecialErrorPageNavigationHandler+SSL.swift]: let's remove empty lines

File: DuckDuckGo/SpecialErrorPage/SpecialErrorPageNavigationHandler+SSL.swift (hunk 3 above)
Comment (jaceklyp) [reply DuckDuckGo/SpecialErrorPage/SpecialErrorPageNavigationHandler+SSL.swift:82]: tbh I don't mind leaving it as it is for now

File: DuckDuckGo/SpecialErrorPage/SpecialErrorPageInterfaces.swift [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    /// Sets the user script for the special error page.
+    func setUserScript(_ userScript: SpecialErrorPageUserScript?)
+}
+
+typealias SpecialErrorPageNavigationHandling = BaseSpecialErrorPageNavigationHandling & WebViewNavigationHandling & SpecialErrorPageUserScriptDelegate
+
+// MARK: - SpecialErrorPageNavigationDelegate
+
+/// A delegate for handling navigation actions related to special error pages.
+protocol SpecialErrorPageNavigationDelegate: AnyObject {
```
Comment (jaceklyp) [inline DuckDuckGo/SpecialErrorPage/SpecialErrorPageInterfaces.swift]: can we move this protocol above BaseSpecialError protocol where it is already used?

File: DuckDuckGo/SpecialErrorPage/SpecialErrorPageInterfaces.swift [hunk 9]
```diff
@@ -0,0 +1,131 @@
+//
+//  SpecialErrorPageInterfaces.swift
```
Comment (jaceklyp) [inline DuckDuckGo/SpecialErrorPage/SpecialErrorPageInterfaces.swift]: I don't think we should put all these protocols under this umbrella file - for me personally it makes things confusing - I'm fine with putting separate files into group though named like this

File: DuckDuckGo/SpecialErrorPage/SpecialErrorPageInterfaces.swift [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+    var failedURL: URL? { get }
+
+    /// Attaches a web view to the special error page handling.
+    func attachWebView(_ webView: WKWebView)
+
+    /// Sets the user script for the special error page.
+    func setUserScript(_ userScript: SpecialErrorPageUserScript?)
+}
+
+typealias SpecialErrorPageNavigationHandling = BaseSpecialErrorPageNavigationHandling & WebViewNavigationHandling & SpecialErrorPageUserScriptDelegate
```
Comment (jaceklyp) [inline DuckDuckGo/SpecialErrorPage/SpecialErrorPageInterfaces.swift]: perhaps we can rename these to:
SpecialErrorPageNavigationHandling  -> SpecialErrorPageManaging
BaseSpecialErrorPageNavigationHandling -> SpecialErrorPageContextHandling/SpecialErrorPageFlowHandling (as it is mostly responsible for state handling and setup)? Base prefix doesn't tell much

File: DuckDuckGoTests/SpecialErrorPage/SSLErrorPageNavigationHandlerTests.swift (hunk 6 above)
Comment (jaceklyp) [reply DuckDuckGoTests/SpecialErrorPage/SSLErrorPageNavigationHandlerTests.swift]: would be great :)

File: DuckDuckGoTests/SpecialErrorPage/SpecialErrorPageNavigationHandlerIntegrationTests.swift [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Testing
+import WebKit
+import SpecialErrorPages
+@testable import DuckDuckGo
+
+@Suite("Special Error Pages - SSL Integration Tests", .serialized)
```
Comment (jaceklyp) [inline DuckDuckGoTests/SpecialErrorPage/SpecialErrorPageNavigationHandlerIntegrationTests.swift:25]: 👏

**Review by jaceklyp — CHANGES_REQUESTED** [review]
what a nice piece of code, you're definitely leaving it in much better shape than it was, the other comments are just minor;
note: haven't tested it yet

**Comment by jaceklyp** [comment]
> what a nice piece of code, you're definitely leaving it in much better shape than it was, the other comments are just minor; note: haven't tested it yet

update: tested and it works well

File: DuckDuckGo/SpecialErrorPage/SpecialErrorPageInterfaces.swift (hunk 9 above)
Comment (alessandroboron) [reply DuckDuckGo/SpecialErrorPage/SpecialErrorPageInterfaces.swift]: Sounds good to me!

**Comment by alessandroboron** [comment]
> what a nice piece of code, you're definitely leaving it in much better shape than it was, the other comments are just minor; note: haven't tested it yet

Thanks @jaceklyp, I will address the comments! I will re-request a review soon :)

File: DuckDuckGo/SpecialErrorPage/SpecialErrorPageInterfaces.swift (hunk 10 above)
Comment (alessandroboron) [reply DuckDuckGo/SpecialErrorPage/SpecialErrorPageInterfaces.swift]: Yes I wasn’t fan of “Base” as well but couldn’t find a better name. :)

File: DuckDuckGoTests/SpecialErrorPage/SSLErrorPageNavigationHandlerTests.swift (hunk 6 above)
Comment (alessandroboron) [reply DuckDuckGoTests/SpecialErrorPage/SSLErrorPageNavigationHandlerTests.swift]: I update the existing tests to use Swift Testing

**Review by jaceklyp — APPROVED** [review]

**Pull request merged**